src/sentry/services/hybrid_cloud/rpc/endpoints.py

```diff
@@ -0,0 +1,236 @@
+            except TypeSerializationError as e:
+                raise TypeSerializationError(
+                    f"Service {self.service_name!r} has invalid published method {k!r}: {e}"
+                )
+
+        return result
+
+    def _is_published_method(self, v: Any):
+        if not inspect.ismethod(v):
+            return False
+        if not getattr(v, "__isabstractmethod__"):
+            return False
+        return True
+
+
+@dataclasses.dataclass
+class MethodEndpoint:
+    params_serializer: JsonSerializer
+    result_serializer: JsonSerializer
+    method_name: str
+
+
+endpoint_registry: MutableMapping[str, ServiceEndpoint] = {}
+
+
+def expose_as_region_silo_rpc(
+    service_name: str, f: Callable[[], ServiceInterface]
+) -> Callable[[], ServiceInterface]:
+    if service_name in endpoint_registry:
+        raise ValueError(f"Service {service_name} registered twice!")
+
+    endpoint_registry[service_name] = ServiceEndpoint(
+        factory=f, service_name=service_name, silo_mode=SiloMode.REGION
+    )
+    return f
+
+
+def impl_with_region_silo_client(
+    service_name: str, service_base: Type[ServiceInterface]
+) -> Callable[[], ServiceInterface]:
+    pass
+
+
+class RpcClient(abc.ABC):
+    @abc.abstractmethod
+    def route(self, method_name: str, parameters: Mapping[str, Any]) -> BaseSiloClient:
+        pass
+
+    def invoke(
+        self, method_name: str, parameters: Mapping[str, Any], result_dataclass: Type[_T]
+    ) -> _T:
+        pass
+
+
+class RegionRpcClient(Generic[ServiceInterface], RpcClient):
+    service_name: str
+    service_base: ServiceInterface
+
+    def __init__(self, service_name: str, service_base: ServiceInterface):
+        self.service_name = service_name
+        self.service_base = service_base
+
+    def route(self, method_name: str, parameters: Mapping[str, Any]) -> BaseSiloClient:
+        if not hasattr(self.service_base, method_name):
+            raise NoMethodError(f"No method for rpc {self.service_name} {repr(method_name)}")
+
+        method = getattr(self.service_base, method_name)
+        org_routing: Callable[..., str | int] | None = getattr(method, "__org_routing")
+
+        org: str | int = org_routing(**parameters)
+        region_name: str | None
+        if isinstance(org, str):
+            region_name = OrganizationMapping.find_region_name_by_org_slug(org)
+        else:
+            region_name = OrganizationMapping.find_region_name_by_org_id(org)
+
+        if region_name is None:
+            raise MissingOrgError(
+                f"Organization {repr(org)} does not exist in OrganizationMapping."
+            )
+
+        try:
+            region = get_region_by_name(region_name)
+        except RegionResolutionError as e:
+            raise MissingRegionError(str(e))
+
+        return RegionSiloClient(region)
+
+
+class HasOrganizationId(Protocol):
+    organization_id: int
+
+
+class HasOrganizationSlug(Protocol):
+    organization_slug: str
```

Comment:
Generally speaking the issues related to slug renames are already pretty annoying without all those latency issues, with replication / outbox / RPCs this will get worse. Ideally at least what sends through the outbox never talks slugs but org IDs when it can.

Slug renames on the org level are also awful for auditing reasons. We have a bunch of business metrics that are messed up because people renamed the org and that reporting assumed that you have some pairing up of IDs to slugs.

We also really should consider a cooldown on org slug renames and make this a more scary operation!

Reply:
With all the problems that orgslugs present, we shouldn't have it as as way to route outbox messages. Slugs are not stable identifiers, and having a customer swap a slug with undelivered messages could go *really* wrong. We could add/remove members to the wrong org, or deliver webhooks to the wrong org.

Reply:
Good point.  There are, some special cases we do have to handle, RE: url routes with orgslugs.  But those can be specially cased, and they are only ever *read operations*.

Reply:
Could we remove the ability to route RPC requests by slug? That will insulate this subsystem from slug drift.

Reply:
I would also generally prefer we stick to the ids. If we do have some requests that lack that information, we better do this as an initial step to retrieve that information.